File: frontend/src/types/webrtc.ts
/**
 * WebRTC 시그널링 관련 타입 정의
 */

// 시그널링 메시지 타입
export type SignalingMessageType =
  | 'join'
  | 'offer'
  | 'answer'
  | 'ice-candidate'
  | 'leave'
  | 'mute'
  | 'joined'
  | 'participant-joined'
  | 'participant-left'
  | 'participant-muted'
  | 'meeting-ended'
  | 'error'
  // 녹음 관련
  | 'recording-offer'
  | 'recording-ice'
  | 'recording-stop'
  | 'recording-answer'
  | 'recording-started'
  | 'recording-stopped'
  // 화면공유 관련
  | 'screen-share-start'
  | 'screen-share-stop'
  | 'screen-share-started'
  | 'screen-share-stopped'
  | 'screen-offer'
  | 'screen-answer'
  | 'screen-ice-candidate';

// 회의실 참여자 정보
export interface RoomParticipant {
  userId: string;
  userName: string;
  role: 'host' | 'participant';
  audioMuted: boolean;
  isScreenSharing?: boolean;
}

// ICE 서버 설정
export interface IceServer {
  urls: string;
  username?: string;
  credential?: string;
}

// 회의실 정보 응답
export interface MeetingRoomResponse {
  meetingId: string;
  status: string;
  participants: RoomParticipant[];
  iceServers: IceServer[];
  maxParticipants: number;
}

// 회의 시작 응답
export interface StartMeetingResponse {
  meetingId: string;
  status: string;
  startedAt: string;
}

// 회의 종료 응답
export interface EndMeetingResponse {
  meetingId: string;
  status: string;
  endedAt: string;
}

// ===== Client -> Server 메시지 =====

export interface JoinMessage {
  type: 'join';
}

export interface OfferMessage {
  type: 'offer';
  sdp: RTCSessionDescriptionInit;
  targetUserId: string;
}

export interface AnswerMessage {
  type: 'answer';
  sdp: RTCSessionDescriptionInit;
  targetUserId: string;
}

export interface IceCandidateMessage {
  type: 'ice-candidate';
  candidate: RTCIceCandidateInit;
  targetUserId?: string;
}

export interface LeaveMessage {
  type: 'leave';
}

export interface MuteMessage {
  type: 'mute';
  muted: boolean;
}

// 녹음 관련 (Client -> Server)
export interface RecordingOfferMessage {
  type: 'recording-offer';
  sdp: RTCSessionDescriptionInit;
}

export interface RecordingIceMessage {
  type: 'recording-ice';
  candidate: RTCIceCandidateInit;
}

export interface RecordingStopMessage {
  type: 'recording-stop';
}

// 화면공유 관련 (Client -> Server)
export interface ScreenShareStartMessage {
  type: 'screen-share-start';
}

export interface ScreenShareStopMessage {
  type: 'screen-share-stop';
}

export interface ScreenOfferMessage {
  type: 'screen-offer';
  sdp: RTCSessionDescriptionInit;
  targetUserId: string;
}

export interface ScreenAnswerMessage {
  type: 'screen-answer';
  sdp: RTCSessionDescriptionInit;
  targetUserId: string;
}

export interface ScreenIceCandidateMessage {
  type: 'screen-ice-candidate';
  candidate: RTCIceCandidateInit;
  targetUserId: string;
}

export type ClientMessage =
  | JoinMessage
  | OfferMessage
  | AnswerMessage
  | IceCandidateMessage
  | LeaveMessage
  | MuteMessage
  | RecordingOfferMessage
  | RecordingIceMessage
  | RecordingStopMessage
  | ScreenShareStartMessage
  | ScreenShareStopMessage
  | ScreenOfferMessage
  | ScreenAnswerMessage
  | ScreenIceCandidateMessage;

// ===== Server -> Client 메시지 =====

export interface JoinedMessage {
  type: 'joined';
  participants: RoomParticipant[];
}

export interface ParticipantJoinedMessage {
  type: 'participant-joined';
  participant: RoomParticipant;
}

export interface ParticipantLeftMessage {
  type: 'participant-left';
  userId: string;
}

export interface ServerOfferMessage {
  type: 'offer';
  sdp: RTCSessionDescriptionInit;
  fromUserId: string;
}

export interface ServerAnswerMessage {
  type: 'answer';
  sdp: RTCSessionDescriptionInit;
  fromUserId: string;
}

export interface ServerIceCandidateMessage {
  type: 'ice-candidate';
  candidate: RTCIceCandidateInit;
  fromUserId: string;
}

export interface ParticipantMutedMessage {
  type: 'participant-muted';
  userId: string;
  muted: boolean;
}

export interface MeetingEndedMessage {
  type: 'meeting-ended';
  reason: string;
}

export interface ErrorMessage {
  type: 'error';
  code: string;
  message: string;
}

// 녹음 관련 (Server -> Client)
export interface RecordingAnswerMessage {
  type: 'recording-answer';
  sdp: RTCSessionDescriptionInit;
}

export interface RecordingStartedMessage {
  type: 'recording-started';
  userId: string;
}

export interface RecordingStoppedMessage {
  type: 'recording-stopped';
  userId: string;
}

// 화면공유 관련 (Server -> Client)
export interface ScreenShareStartedMessage {
  type: 'screen-share-started';
  userId: string;
}

export interface ScreenShareStoppedMessage {
  type: 'screen-share-stopped';
  userId: string;
}

export interface ServerScreenOfferMessage {
  type: 'screen-offer';
  sdp: RTCSessionDescriptionInit;
  fromUserId: string;
}

export interface ServerScreenAnswerMessage {
  type: 'screen-answer';
  sdp: RTCSessionDescriptionInit;
  fromUserId: string;
}

export interface ServerScreenIceCandidateMessage {
  type: 'screen-ice-candidate';
  candidate: RTCIceCandidateInit;
  fromUserId: string;
}

export type ServerMessage =
  | JoinedMessage
  | ParticipantJoinedMessage
  | ParticipantLeftMessage
  | ServerOfferMessage
  | ServerAnswerMessage
  | ServerIceCandidateMessage
  | ParticipantMutedMessage
  | MeetingEndedMessage
  | ErrorMessage
  | RecordingAnswerMessage
  | RecordingStartedMessage
  | RecordingStoppedMessage
  | ScreenShareStartedMessage
  | ScreenShareStoppedMessage
  | ServerScreenOfferMessage
  | ServerScreenAnswerMessage
  | ServerScreenIceCandidateMessage;

// 연결 상태
export type ConnectionState = 'disconnected' | 'connecting' | 'connected' | 'reconnecting' | 'failed';

// 피어 연결 정보
export interface PeerConnection {
  peerId: string;
  connection: RTCPeerConnection;
  stream?: MediaStream;
}

File: backend/app/schemas/webrtc.py
"""LiveKit 기반 회의 관련 Pydantic 스키마"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class RoomParticipant(BaseModel):
    """회의실 참여자 정보"""
    user_id: UUID = Field(serialization_alias="userId")
    user_name: str = Field(serialization_alias="userName")
    role: str
    audio_muted: bool = Field(default=False, serialization_alias="audioMuted")

    class Config:
        populate_by_name = True


class StartMeetingResponse(BaseModel):
    """회의 시작 응답"""
    meeting_id: UUID = Field(serialization_alias="meetingId")
    status: str
    started_at: datetime = Field(serialization_alias="startedAt")

    class Config:
        populate_by_name = True


class EndMeetingResponse(BaseModel):
    """회의 종료 응답"""
    meeting_id: UUID = Field(serialization_alias="meetingId")
    status: str
    ended_at: datetime = Field(serialization_alias="endedAt")

    class Config:
        populate_by_name = True


# ===== LiveKit 관련 스키마 =====


class LiveKitTokenResponse(BaseModel):
    """LiveKit 토큰 응답"""

    token: str
    ws_url: str = Field(serialization_alias="wsUrl")
    room_name: str = Field(serialization_alias="roomName")

    class Config:
        populate_by_name = True


class LiveKitRoomResponse(BaseModel):
    """LiveKit 룸 정보 응답"""

    meeting_id: UUID = Field(serialization_alias="meetingId")
    room_name: str = Field(serialization_alias="roomName")
    status: str
    participants: list[RoomParticipant]
    max_participants: int = Field(default=20, serialization_alias="maxParticipants")
    ws_url: str = Field(serialization_alias="wsUrl")
    token: str

    class Config:
        populate_by_name = True


class StartRecordingResponse(BaseModel):
    """녹음 시작 응답"""

    meeting_id: UUID = Field(serialization_alias="meetingId")
    egress_id: str = Field(serialization_alias="egressId")
    started_at: datetime = Field(serialization_alias="startedAt")

    class Config:
        populate_by_name = True


class StopRecordingResponse(BaseModel):
    """녹음 중지 응답"""

    meeting_id: UUID = Field(serialization_alias="meetingId")
    stopped_at: datetime = Field(serialization_alias="stoppedAt")

    class Config:
        populate_by_name = True
